src/pg_query_summary.c

```diff
@@ -0,0 +1,761 @@
+		if (list_cell_number(funcname, lc) != 0)
+			appendStringInfoChar(&buf, '.');
+
+		String* str = lfirst_node(String, lc);
+		appendStringInfoString(&buf, quote_identifier(str->sval));
+	}
+
+	SummaryFunction* func = palloc(sizeof(SummaryFunction));
+
+	func->name = buf.data;
+
+	if (list_length(funcname) == 3)
+		func->schema_name = lsecond_node(String, funcname)->sval;
+	else if (list_length(funcname) == 2)
+		func->schema_name = linitial_node(String, funcname)->sval;
+	else
+		func->schema_name = NULL;
+
+	func->function_name = llast_node(String, funcname)->sval;
+
+	func->context = context;
+
+	state->functions = lappend(state->functions, func);
+}
+
+bool pg_query_summary_walk_impl(Node* node, WalkState* state) {
+	if (node == NULL)
+		return false;
+
+	// NOTE: If `raw_expression_tree_walker()` doesn't handle a particular
+	// node type, you will need to `return false` early to avoid passing it
+	// to the tree walker.
+	// If you do this, be sure to update `pg_query_summary_stmt_walk_impl()`.
+
+	switch (nodeTag(node)) {
+		case T_CommonTableExpr: {
+			CommonTableExpr* cte = castNode(CommonTableExpr, node);
+			state->cte_names = lappend(state->cte_names, pstrdup(cte->ctename));
+			break;
+		}
+
+		case T_SelectStmt: {
+			SelectStmt *stmt = castNode(SelectStmt, node);
+
+			pg_query_summary_walk_impl((Node*) stmt->targetList, state);
+
+			if (stmt->op == SETOP_NONE)
+				add_from_clause_items(stmt->fromClause, CONTEXT_SELECT, state);
+
+			// For SETOP_UNION, SETOP_EXCEPT, and SETOP_INTERSECT,
+			// the raw_expression_tree_walker() call at the end of
+			// this function will handle stmt->larg/stmt->rarg.
+
+			if (stmt->intoClause)
+				add_from_clause_item(stmt->intoClause->rel, CONTEXT_DDL, state);
+
+			if (stmt->whereClause) {
+				// If we have a WHERE clause, that means we have to check for filter columns.
+				//
+				// NOTE: the WHERE clause is walked twice, unfortunately.
+				//  I don't know how to avoid this. It doesn't *appear*
+				//  to cause problems, but it means there's definitely an
+				//  optimization opportunity available.
+				state->save_filter_columns = true;
+				pg_query_summary_walk_impl((Node*) stmt->whereClause, state);
+				state->save_filter_columns = false;
+			}
+
+			break;
+		}
+
+		case T_InsertStmt: {
+			add_from_clause_item(castNode(InsertStmt, node)->relation, CONTEXT_DML, state);
+			break;
+		}
+
+		case T_UpdateStmt: {
+			UpdateStmt *stmt = castNode(UpdateStmt, node);
+			add_from_clause_item(stmt->relation, CONTEXT_DML, state);
+			add_from_clause_items(stmt->fromClause, CONTEXT_SELECT, state);
+			break;
+		}
+
+		case T_DeleteStmt: {
+			DeleteStmt *stmt = castNode(DeleteStmt, node);
+			add_from_clause_item(stmt->relation, CONTEXT_DML, state);
+			add_from_clause_items(stmt->usingClause, CONTEXT_SELECT, state);
+			break;
+		}
+
+		case T_CopyStmt: {
+			CopyStmt* stmt = castNode(CopyStmt, node);
+			add_from_clause_item(stmt->relation, CONTEXT_SELECT, state);
```

Comment:
```suggestion
			add_from_clause_item(stmt->relation, CONTEXT_SELECT, state);
			pg_query_summary_walk_impl((Node*) stmt->query, state);
```

Like in https://github.com/pganalyze/pg_query/blob/d8a6b8f2028978c3e9295b4129060fd808bd6989/lib/pg_query/parse.rb#L174C37-L174C52 we need to handle `COPY (SELECT * FROM tbl) TO file.test` type statements.